src/GafferDispatch/Dispatcher.cpp

```diff
@@ -648,6 +648,11 @@ class Dispatcher::Batcher
 			}
 		}
 
+		const Gaffer::Plug *dispatcherPlug( const TaskNode::Task &task )
+		{
+			return static_cast<const TaskNode *>( task.plug()->node() )->dispatcherPlug();
```

Comment:
I had to think a bit about whether this would ever fail.  If you had a node implemented as a network, with a task output, and then you removed the internal task nodes ( maybe the network for some reason no longer needs to submit tasks ), then you could end up with source() returning an output plug on the network, which might not be a TaskNode, causing a crash?  This is an obscure situation, but it means that I'm not super clear on what the rules for TaskPlugs are.  You may only have an output TaskPlug on a TaskNode, or a network ( if it's connected to an internal TaskNode )?

Reply:
GafferDispatch used to be pretty hazy in this area, mixing and matching where it passed plugs vs nodes, and I'm slowly refactoring it into line with ComputeNode/ValuePlug. The general rule in all this is that nodes are the "doers" but their API is protected, and plugs provide the public API. You don't call `ComputeNode::compute()`, you call `ValuePlug::getValue()`, and you don't call `TaskNode::execute()`, you call `TaskPlug::execute()`. The rule for output TaskPlugs is also analogous to that for output ValuePlugs : the source plug needs to belong to a TaskNode. If you inadvertently build an invalid network, this will trigger the [exception in TaskNodeProcess](https://github.com/GafferHQ/gaffer/blob/master/src/GafferDispatch/TaskNode.cpp#L111). This will be triggered before we call `dispatcherPlug()`, so the cast here is safe. The prior code used `runTimeCast()` via the removed `Task::node()` method, but did not check for failure, making `runTimeCast()` pointless. Since we have checked that we have a TaskNode already, I felt that `static_cast` was appropriate here.

Reply:
I figured that this was subtle enough as to warrant some test coverage, so I've pushed [one further commit](0a83d06), which adds some. I'll merge once the CI passes.